Core/Player: implement Stage 1 of WarMode

matanshukry opened the pull request:
**Changes proposed:**

-  Warmode

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

Closes #22734
Closes #25918

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [x] Tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

- [x] this is really weird, but it seems the `UNIT_FLAG_PVP_ATTACKABLE` flag is flipped somehow. When it's added to the player, he isn't attackable (and can't attack). When it's removed, he is attackable and can attack.
- [x] 2nd stage would be to separate war mode into phases. That is, each war mode should have it's own phase, which is pretty much like an entire world (perhaps actually 2 instances of `World` class? idk). *Edit*: created separate issue for it #25975
- [x] Grant extra experience/money based on enlisted aura
- [x] `WarMode: Calls to Arms` buff added when faction is outnumbered.
- [x] Persist `WarMode: Calls to Arms` calculated data.
- [x] `Against Overwhelming Odds` quest giving if a faction is overwhelmingly outnumbered - *Edit*: created separate issue for it #26002
- [x] Bounty Hunter System *Edit*: created separate issue for it #25974

matanshukry committed:
Implemented 1st step of warmode

matanshukry committed:
Give warmode extra xp on killing creature or xp, and extra money if at
max level

osleyder85 commented:
War Mode Requirements:

1. When the player reaches level 20, the PvP talents window must be activated.

2. Players may only activate War Mode at Stormwind City or Orgrimmar by using the Talents interface, or deactivate War Mode at any rested location (including inns and other cities). 

3. War Mode replaced previous open-world PvP concepts which had endured from Classic until Legion:
    - PvP realms with players always targetable by the opposing faction.
    - PvE realms where players could opt in/out in the span of five minutes from anywhere in the world.

4. Apply the spell to the player as appropriate:
    - [Enlisted](https://www.wowhead.com/spell=282559/enlisted)
    - [Enlisted](https://www.wowhead.com/spell=289954/enlisted)
    - [Enlisted](https://www.wowhead.com/spell=269083/enlisted)

5. With War Mode enabled, a +10% bonus applies to experience from quests and mob kills, and to gold, currency or Azerite Power from world quests. If a faction is outnumbered, a War Mode: Call to Arms adds up to a further +20% bonus. Furthermore, if the participation is overwhelmingly one-sided, then the opposing faction will be offered, as even further incentive to participate:
    
    - [Against Overwhelming Odds](https://www.wowhead.com/quest=54979/against-overwhelming-odds)

6.  Bounty Hunter System:
Killing a certain number of players of the opposite faction, without dying awards the status of Assassin that increases damage and healing by 15%. The downside to this, the more players an Assassin kills, the more likely they will show up on the zone map for other players to target and to gain a bounty. Players who successfully kill an Assassin of the opposite faction will receive Conquest points as well as other rewards. There can be three or possibly more Assassins on Zandalar or Kul Tiras at a time, providing plenty of opportunity for players or groups of players to take their revenge on the opposite faction's Assassins.

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -15760,15 +15760,20 @@ void Player::RewardQuest(Quest const* quest, LootItemType rewardType, uint32 rew
+    float moneyMod = 1.0f;
     if (getLevel() < sWorld->getIntConfig(CONFIG_MAX_PLAYER_LEVEL))
-        GiveXP(XP, nullptr);
+        GiveXP(XP * (HasWarmodeEnlistedAura() ? GetWarmodeAuraMod() : 1.0f), nullptr);
```
Unneccessary - 269083 has SPELL_AURA_MOD_XP_QUEST_PCT, implemented in `Player::GetQuestXPReward`

matanshukry replied:
It's not working. But I guess that's another bug.

Shauren replied:
Yes

matanshukry commented:
@osleyder85 I've updated my post with further todo list based on your message.
Question for clarification: is the increase also for currency, and not just for money? I see it written on gamepedia.com for example, but the spells description have no mention of currency.

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -15760,15 +15760,20 @@ void Player::RewardQuest(Quest const* quest, LootItemType rewardType, uint32 rew
     else
+    {
         moneyRew = int32(quest->GetRewMoneyMaxLevel() * sWorld->getRate(RATE_DROP_MONEY));
+        moneyMod = GetWarmodeAuraMod();
```
Should be implemented using SPELL_AURA_MOD_MONEY_GAIN inside `Player::GetQuestMoneyReward` (with `GetTotalAuraModifierByMiscValue(SPELL_AURA_MOD_MONEY_GAIN, 0)`)

Shauren commented on src/server/game/Miscellaneous/Formulas.h:
```diff
@@ -201,6 +201,10 @@ namespace Trinity
                 gain = uint32(gain * xpMod);
+
+                // Warmode buff
+                if (player->HasWarmodeEnlistedAura())
```
Also unneccessary, 269083 has SPELL_AURA_MOD_XP_PCT

Shauren commented:
All bonuses are handled using that aura, don't add any special code for it

osleyder85 commented:
SpellID: 289954 - Enlisted
- Artifact Power, War Resources, and Gold rewards from world quests increased.
- Experience gained from killing monsters and completing quests increased by 10%.

SpellID: 282559 - Enlisted
- Anima and Gold
- Rewards from world quests increased by 10%.
- Experience gained from killing monsters and completing quests increased by 0%.

matanshukry committed:
- Removed extra xp & currency. They're automatically handled with
  effects & auras
- Using different aura if player is at max level

MaxtorCoder commented:
Bouty Hunter system requires the Vignette system to be implemented. You could create a PR for that separately.

Mesielecat commented:
The bounty hunter is only a feature that require War Mode, its not necesary for this PR, and as Maxtor said, Bounty Hunter require Vignette System and the implementation of the SPELL_AURA_266 (that could be renamed to SPELL_AURA_VIGNETTE), and the opcode SMSG_VIGNETTE_UPDATE. Vignette and Bounty Hunter need different PR otherwise going to need lot of review and going to take long of time to be merged.

matanshukry commented:
@MaxtorCoder @Mesielecat I've created a separate PR for the bounty hunter system.

Regarding what's left on this PR:
- `UNIT_FLAG_PVP_ATTACKABLE` is flipped, which is weird. Can anyone comment about it?
- WarMode: Calls to Arms buff added when faction is outnumbered. Can anyone give an estimation on how much is outnumbered?
- Against Overwhelming Odds quest giving if a faction is overwhelmingly outnumbered. Can anyone give an estimation here too?

p.s. I'm assuming the estimations are in %.

mdX7 commented:
> 
> 
> @MaxtorCoder @Mesielecat I've created a separate PR for the bounty hunter system. Thank you both for the information!
> 
> Regarding what's left on this PR:
> 
>     * `UNIT_FLAG_PVP_ATTACKABLE` is flipped, which is weird. Can anyone comment about it?
> 
>     * WarMode: Calls to Arms buff added when faction is outnumbered. Can anyone give an estimation on how much is outnumbered?
> 
>     * Against Overwhelming Odds quest giving if a faction is overwhelmingly outnumbered. Can anyone give an estimation here too?
> 
> 
> p.s. I'm assuming the estimations are in %.

Sadly Blizzard is not giving is any option to request faction balance anymore. 

I tried an approach, I calculated it via. https://www.wowprogress.com/realms/rank/eu (and us)

EU: Horde: 333216 (60.54091766321825%), Alliance: 217182 (39.45908233678175%)
US: Horde: 265414 (59.34234973471747%), Alliance: 181845 (40.65765026528253%)

In EU Against Overwhelming Odds is available for Alliance players so i guess its the same way for US. (iirc it is region based on retail, not realm based)

I retrieved it using this script i shitted together on the fly.
```py
#!/bin/python3

import requests
import re

def main():
    r = requests.get('https://www.wowprogress.com/realms/rank/eu')
    hordeList = [int(i) for i in re.findall(r'<span class=\"num horde\">([0-9]+)<', r.text)]
    allianceList = [int(i) for i in re.findall(r'<span class=\"num alliance\">([0-9]+)<', r.text)]
    hordePlayers = sum(hordeList)
    alliancePlayers = sum(allianceList)
    overallPlayers = hordePlayers + alliancePlayers
    print('Horde: {} ({}%), Alliance: {} ({}%)'.format(hordePlayers, (hordePlayers/overallPlayers*100), alliancePlayers, (alliancePlayers/overallPlayers*100)))

if __name__ == '__main__':
    main()
```

illfated commented:
BTW, PR title suggestion (use what you want from it, or leave it to the dev who merges it):

**Core/Player: implement Stage 1 of WarMode**

matanshukry committed:
Warmode value is updated based on factions count

matanshukry commented:
Update:
- Haven't received any comments on `UNIT_FLAG_PVP_ATTACKABLE`, so I'm going to assume it's fine.
- Implemented `WarMode: Calls to Arms` buff value; +5/10/15% on the enlisted buff based on value. Include debug commands for gm.

Missing:
- From online source I'm assuming the Enlisted buff should be calculated weekly, and kept. The weekly reset sounds like a good time. Correct me if I'm wrong.
- I would like to persist the calculated enlisted value, but not sure if there's a db table for that or I should create one.

mdX7 commented on src/server/game/World/World.cpp:
```diff
@@ -3684,6 +3747,98 @@ void World::RemoveOldCorpses()
+    switch (reward)
+    {
+        case FactionOutnumberReward::Overwhelming:
+        case FactionOutnumberReward::Percent20: return baseValue + 20;
```
where does this 20% come from? "overwhelming" is equal to 5% on retail currently.
More specific (EU):
Horde: 10% (normal)
Alliance: 15% (has overwhelming odds quest)

matanshukry replied:
From various sources online. The source can be *up* to 20%, depending on the amount.
For example:
https://wow.gamepedia.com/War_Mode ` If a faction is outnumbered, a War Mode: Call to Arms adds up to a further +20% bonus.`

https://www.reddit.com/r/wow/comments/kq0nj5/war_mode_enlisted_bonuses/gi34b4f/?utm_source=reddit&utm_medium=web2x&context=3
I guess it's a lot more rare since the percentage is higher, but I believe this is how it's done.

You disagree?

Shauren replied:
That wiki has wrong info. Call to Arms bonus scales in steps of 5%

matanshukry replied:
@Shauren I don't think it says otherwise, and that's in fact what I think it is. Although I don't think a total of 25% bonus exist; I haven't seen one anyway.

@mdX7 I believe I misread what you wrote.
At first I thought you meant there shouldn't be any 20% extra, and the maximum should be 10% (total 20%)).
Reading again, I now understand it means:
1. 5% extra (15% total)
2. 10% extra (20% total) + overwhelming
3. 20% extra (30% total)

The part about overwhelming being 20% and not 10% was a complete guess and not from a soruce.
So yeah I will change it.

matanshukry committed:
Removed unused variable

matanshukry committed:
Removed overwhelming as a config variable. Should simply be rewareded
when giving 10% or 20% bonus to enlisted

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -28727,3 +28731,74 @@ uint8 Player::GetItemLimitCategoryQuantity(ItemLimitCategoryEntry const* limitEn
+            CustomSpellValues const* values = nullptr;
+            if (GetTeamId() == team)
+                values = &(sWorld->GetCurrentFactionBalanceRewardSpellValues());
+            CastSpell(this, IsAtMaxLevel() ? auraOutsideMaxLvl : auraOutside, true, nullptr, nullptr, ObjectGuid::Empty, values);
```
Use `void Unit::CastCustomSpell(uint32 spellId, CustomSpellValues const& value, Unit* victim, TriggerCastFlags triggerFlags, Item* castItem, AuraEffect const* triggeredByAura, ObjectGuid originalCaster)`

matanshukry replied:
Done.

There is a `CastSpell` override that accepts `CustomSpellValues` as one of it's argument; should it be renamed to CastCustomSpell ?

matanshukry committed:
Using CastCustomSpell when relevant rather than editing the CastSpell
overrides

matanshukry commented:
I removed the overwhelming quest part from this issue. It requires the world quest feature, which isn't implemented yet and require further investigation into the part in the quest log that's after the 25 initial slots.

I created 2 issues for them; 1 for the overwhelming odds quest, and 1 for the world quests feature.

matanshukry commented:
Currently the only issue that's left is
> Persist WarMode: Calls to Arms calculated data.`

Any thoughts on how to do that? is there a specific table I can use for that? do I need to create a new table?

Shauren commented:
* UNIT_FLAG_PVP_ATTACKABLE - this must be always set, no exception. Silly side effect of not having it: if you enter water client will ignore water and behave as if it was not there
* Revert the change to SPELL_AURA_PVP_TALENTS as old code is correct - pvp talents are not tied to war mode (example spell 244983, used on bfa island expeditions feature in pvp mode)
* Where did spell 289954 come from? It is not used on retail
* Places in code that currently check `sWorld->IsPvPRealm()` should be replaced with checking `IsWarModeActive()` on player
* UNIT_FLAG2_ALLOW_CHANGING_TALENTS should not be checked to allow disabling war mode
* Disallow changing war mode status when player has aura SPELL_PVP_RULES_ENABLED
* Instead of hardcoded area ids, the correct check is `(factionTemplate->FriendGroup & area->FactionGroupMask) && (area->Flags[1] & AREA_FLAG_2_CAN_ENABLE_WAR_MODE)` (see Player::UpdateHostileAreaState) - flag discovered & pushed in 086f337d1534be5b332670fb41a8b883ab731a3a. Maybe wrap that logic in a function (CanEnableWarModeInArea) and use that to also determine inside/outside buff id
* For the lulz i visited stormwind on my horde character and PLAYER_FLAGS_WAR_MODE_ACTIVE was not applied (but i had the "outside" buff) - this seems to indicate that setting this flag is actually tied to the server/instance you are on, not actual war mode state (you also cannot see warmode exclusive world quests on such servers, even if your war mode is desired)

What you were probably after was `AddPvpFlag(UNIT_BYTE2_FLAG_PVP);`

I have not looked at faction balancing stuff yet, will do that in a later review

matanshukry committed:
Updates after cr

matanshukry commented:
> Instead of hardcoded area ids, the correct check is (factionTemplate->FriendGroup & area->FactionGroupMask) && (area->Flags[1] & AREA_FLAG_2_CAN_ENABLE_WAR_MODE) (see Player::UpdateHostileAreaState) - flag discovered & pushed in 086f337. Maybe wrap that logic in a function (CanEnableWarModeInArea) and use that to also determine inside/outside buff id

isn't `(factionTemplate->FriendGroup & area->FactionGroupMask)` used as a friendly area? e.g. will return true for ironforge with humans ?


> Where did spell 289954 come from? It is not used on retail
From internet sources.
```
Enlisted
Experience gained from killing monsters and completing quests increased by 10%.

Max level: Artifact Power, War Resources, and Gold rewards from world quests increased by 10%.
```

That is, you sholdn't be getting bonus artifact power war resources or gold rewards unless you're at max level.
Looking at it again I think I flipped between them; that is, 289954 should be used while below max level, and 269083 should be used at max level.

Or is that not how it works anymore?

Shauren commented:
> isn't `(factionTemplate->FriendGroup & area->FactionGroupMask)` used as a friendly area? e.g. will return true for ironforge with humans ?

That is why you use it together with AREA_FLAG_2_CAN_ENABLE_WAR_MODE (only stormwind, orgrimmar and their subzones have that)

> 
> ```
> Enlisted
> Experience gained from killing monsters and completing quests increased by 10%.
> 
> Max level: Artifact Power, War Resources, and Gold rewards from world quests increased by 10%.
> ```
> 
> That is, you sholdn't be getting bonus artifact power war resources or gold rewards unless you're at max level.
> Looking at it again I think I flipped between them; that is, 289954 should be used while below max level, and 269083 should be used at max level.
> 
> Or is that not how it works anymore?

Nope, those parts are hidden in tooltip using ?ml (maxlevel) conditional clientside, 269083 is the only one used at all levels

matanshukry commented:
@Shauren
> Places in code that currently check sWorld->IsPvPRealm() should be replaced with checking IsWarModeActive() on player
Currently `IsWarModeActive` returns the state of `PLAYER_FLAGS_WAR_MODE_ACTIVE`.
I also have another function `CanFightOtherFaction` which returns the state of `UNIT_BYTE2_FLAG_PVP`.

Since war mode isn't active in stormwind but fighting other faction is allowed there with wm on - I should probably use `CanFightOtherFaction` rather than `IsWarModeActive` instead of all the `IsPVPRealm`; right?

matanshukry committed:
Added local warmode

illfated commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -7138,8 +7138,8 @@ void Player::UpdateArea(uint32 newArea)
 
-    uint32 const areaRestFlag = (GetTeam() == ALLIANCE) ? AREA_FLAG_REST_ZONE_ALLIANCE : AREA_FLAG_REST_ZONE_HORDE;
-    if (area && area->Flags[0] & areaRestFlag)
+    // uint32 const areaRestFlag = (GetTeam() == ALLIANCE) ? AREA_FLAG_REST_ZONE_ALLIANCE : AREA_FLAG_REST_ZONE_HORDE;
```
I presume commenting out this line is temporary, maybe waiting to be included in a future commit (unless you didn't know that TC usually prefers to have lines removed instead of commented out).

matanshukry replied:
I was being lazy since Shauren wanted to comment on some faction balance things anyway, so I figured I'd just remove it after that.
I'll just do it now though

illfated replied:
Fair enough. Just as long as there are plans to change it later, that is OK too.

matanshukry committed:
Removed commented code

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -28832,3 +28835,114 @@ uint8 Player::GetItemLimitCategoryQuantity(ItemLimitCategoryEntry const* limitEn
+    UpdateWarModeAuras();
+}
+
+bool Player::IsInFactionFriendlyArea(AreaTableEntry const* inArea /* = nullptr */, AreaTableEntry const** outArea /* = nullptr */) const
```
I don't really like this out arg, it would be better to make two separate overloads
```
bool IsInFactionFriendlyArea() const; // this one will just be if (AreaTableEntry const* areaEntry = sAreaTableStore.LookupEntry(GetAreaId()))  IsInFactionFriendlyArea(areaEntry)
bool IsInFactionFriendlyArea(AreaTableEntry const* areaEntry) const;
```

Shauren commented on src/server/game/Entities/Player/Player.h:
```diff
@@ -2548,6 +2552,18 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
+        void SetWarModeDesired(bool enabled);
+        bool IsWarModeDesired() const { return HasPlayerFlag(PLAYER_FLAGS_WAR_MODE_DESIRED); }
+        bool IsWarModeActive() const { return HasPlayerFlag(PLAYER_FLAGS_WAR_MODE_ACTIVE); }
+        bool IsWarModeLocalActive() const { return HasPvpFlag(UNIT_BYTE2_FLAG_PVP); }
```
Wrong flag checked here, should be PLAYER_LOCAL_FLAG_WAR_MODE

Shauren commented on src/server/game/Handlers/CharacterHandler.cpp:
```diff
@@ -1456,6 +1456,11 @@ void WorldSession::HandleTutorialFlag(WorldPackets::Misc::TutorialSetFlag& packe
     }
 }
 
+void WorldSession::HandleSetWarMode(WorldPackets::Character::SetWarMode& packet)
```
Move this to MiscHandler.cpp, next to WorldSession::HandleSetPvP

And you should also check PLAYER_LOCAL_FLAG_WAR_MODE inside both `WorldSession::HandleTogglePvP` and `WorldSession::HandleSetPvP` and return without doing anything - don't want to trigger the 5 minute pvp timer rule when war mode is enabled

matanshukry committed:
- Refactored out arg to 2 methods
- Fixed checking correct flag in IsWarModeLocalActive
- Checking war mode before triggering 5 minute pvp timer

matanshukry commented:
@Shauren ?

Naddley commented:
@matanshukry  
Conflicting files
src/server/game/Entities/Unit/Unit.h
src/server/game/World/World.cpp

matanshukry commented:
@Naddley this PR was waiting on hold before the conflict, I'm guessing because there's some change needed.

I'll just resolve the conflicts together with that change then, once we'll know what it is.

Shauren committed:
move packet to other namespace

Shauren committed:
cleanup player

Shauren committed:
Convert query to prepared statement and aura value calculation to script

Shauren committed:
correct "inside" aura application logic

Shauren committed:
remove unneccessary changes

Shauren committed:
show war mode bonus value in talent ui

Shauren committed:
Update worldserver.conf.dist
